### PR TITLE
BUG 1815039: Retrieve the latest status of SriovNetworkNodeState CR when update is…

### PR DESCRIPTION
**Pull request by pliurh**
… triggered

### CODE REVIEW

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -281,9 +281,24 @@ func (dn *Daemon) nodeStateChangeHandler(old, new interface{}) {
...
+	if len(dn.LoadedPlugins) == 0 {
+		err := dn.loadVendorPlugins(newState)
+		if err != nil {
+			glog.Errorf("nodeStateChangeHandler(): failed to load vendor plugin: %v", err)
+			dn.exitCh <- err
+			return
+		}
+	}
+	// Get the latest NodeState
+	latestState, err := dn.client.SriovnetworkV1().SriovNetworkNodeStates(namespace).Get(dn.name, metav1.GetOptions{})
```
Comment (zshi-redhat) [inline pkg/daemon/daemon.go:293]: Can we guarantee `latestState` is the latest? I think there is a chance that daemon writer not updating node state immediately after policy is deleted.

File: pkg/daemon/daemon.go [hunk 2]
```diff
@@ -281,9 +281,24 @@ func (dn *Daemon) nodeStateChangeHandler(old, new interface{}) {
 	}
 	reqReboot := false
 	reqDrain := false
+	if len(dn.LoadedPlugins) == 0 {
```
[the remark this answers is not in the document]
Comment (pliurh) [reply pkg/daemon/daemon.go:284]: This is to ensure that the plugins will only be loaded once. Normally, the plugins will be loaded when the daemon starts, and `nodeStateAddHandler` will invoke `nodeStateChangeHandler` for the first time.

**Comment by pliurh** [comment]
/hold
I need to refactor this PR a bit.

File: pkg/daemon/daemon.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pliurh) [reply pkg/daemon/daemon.go]: I removed the logic at the latest commit.

File: pkg/daemon/daemon.go (hunk 1 above)
Comment (pliurh) [reply pkg/daemon/daemon.go]: All the CR change event will be put in the workqueue, and be processed one after another. So I add a logic to check if the spec of the newState is identical with the latestState, if not, the event will be skipped. It makes sure that only the event latestState will be processed.

File: pkg/daemon/daemon.go [hunk 3]
```diff
@@ -270,8 +270,20 @@ func (dn *Daemon) nodeStateChangeHandler(old, new interface{}) {
 	var err error
 	newState := new.(*sriovnetworkv1.SriovNetworkNodeState)
 	oldState := old.(*sriovnetworkv1.SriovNetworkNodeState)
-	glog.V(2).Infof("nodeStateChangeHandler(): current generation is %d", newState.GetObjectMeta().GetGeneration())
-	if newState.GetObjectMeta().GetGeneration() == oldState.GetObjectMeta().GetGeneration() {
+	glog.V(2).Infof("nodeStateChangeHandler(): new generation is %d", newState.GetObjectMeta().GetGeneration())
+	// Get the latest NodeState
+	latestState, err := dn.client.SriovnetworkV1().SriovNetworkNodeStates(namespace).Get(dn.name, metav1.GetOptions{})
```
Comment (zshi-redhat) [inline pkg/daemon/daemon.go:275]: shall it use retry here to avoid transient failure (as the error handling is exiting the daemon)?

Comment (pliurh) [reply pkg/daemon/daemon.go]: fixed

**Pull request merged**